src/testRunner/unittests/evaluation/usingDeclarations.ts

```diff
@@ -0,0 +1,1603 @@
+        }
+        `, { target: ts.ScriptTarget.ES2018 });
+
+        main();
+
+        assert.deepEqual(output, [
+            "before block",
+            "enter block",
+            "body",
+            "exit block",
+            "after block"
+        ]);
+    });
+
+    it("'using' for 'undefined' value", () => {
+        const { main, output } = evaluator.evaluateTypeScript(`
+        export const output: any[] = [];
+
+        function body() {
+            output.push("body");
+        }
+
+        export function main() {
+            output.push("before block");
+            {
+                output.push("enter block");
+                using _ = undefined;
+                body();
+                output.push("exit block");
+            }
+            output.push("after block");
+        }
+        `, { target: ts.ScriptTarget.ES2018 });
+
+        main();
+
+        assert.deepEqual(output, [
+            "before block",
+            "enter block",
+            "body",
+            "exit block",
+            "after block"
+        ]);
+    });
+
+    it("'using' for non-disposable value", () => {
+        const { main, output } = evaluator.evaluateTypeScript(`
+        export const output: any[] = [];
+
+        function body() {
+            output.push("body");
+        }
+
+        export function main() {
+            output.push("before block");
+            {
+                output.push("enter block");
+                using _ = {} as any;
+                body();
+                output.push("exit block");
+            }
+            output.push("after block");
+        }
+        `, { target: ts.ScriptTarget.ES2018 });
+
+        assert.throws(main);
+        assert.deepEqual(output, [
+            "before block",
+            "enter block",
+        ]);
+    });
+
+    it("'using' disposes in reverse order", () => {
+        const { main, output } = evaluator.evaluateTypeScript(`
+        export const output: any[] = [];
+
+        const disposable_1 = {
+            [Symbol.dispose]() {
+                output.push("disposed 1");
+            }
+        };
+        const disposable_2 = {
+            [Symbol.dispose]() {
+                output.push("disposed 2");
+            }
+        };
+
+        function body() {
+            output.push("body");
+        }
+
+        export function main() {
+            output.push("before block");
+            {
+                output.push("enter block");
+                using _ = disposable_1, __ = disposable_2;
+                body();
+                output.push("exit block");
+            }
+            output.push("after block");
+        }
+        `, { target: ts.ScriptTarget.ES2018 });
+
+        main();
+
+        assert.deepEqual(output, [
+            "before block",
+            "enter block",
+            "body",
+            "exit block",
+            "disposed 2",
+            "disposed 1",
+            "after block"
+        ]);
+    });

```

Comment:
Predictable how? `using` itself is synchronous disposal only. Are you talking about `await using`? If so, I can use additional promises to block/release two "parallel" async functions if that's what you had in mind. It would be helpful to know what behavior you think needs to be tested.

Reply:
In that it produces a stable list of output, or else an exception. I realised that I was talking about using a Promise of a disposable thing, though, which is not disposable and will give a straightforward compile error. I guess something could implement both Promise and Disposable, but that's not worth testing probably.

Reply:
`await using x = y` doesn't await `y`, so even if `y` is `Promise & Disposable` (or `Promise & AsyncDisposable`) it wouldn't matter as the `await using` is only looking at the disposable part. The `await` part of `await using` is only applied to the result of calling `[Symbol.asyncDispose]()` at the end of the block.